File: chapters/02_channel_adapter_registry/v3_registry_dispatch.py
from __future__ import annotations

from pathlib import Path


class AdapterRegistry:
    def __init__(self, workspace: Path) -> None:
        self.workspace = workspace.resolve()
        self.workspace.mkdir(parents=True, exist_ok=True)
        self._handlers: dict[str, callable] = {}

    def register(self, name: str, handler) -> None:
        self._handlers[name] = handler

    def dispatch(self, name: str, **kwargs) -> str:
        handler = self._handlers.get(name)
        if handler is None:
            return f"unknown adapter:{name}"
        return str(handler(**kwargs))


def _safe_join(base: Path, rel_path: str) -> Path:
    base_resolved = base.resolve()
    candidate = (base_resolved / rel_path).resolve()
    if not str(candidate).startswith(str(base_resolved)):
        raise ValueError("path escape rejected")
    return candidate


def build_registry(workspace: Path) -> AdapterRegistry:
    registry = AdapterRegistry(workspace)

    def write_message(path: str, content: str) -> str:
        p = _safe_join(workspace, path)
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(content, encoding="utf-8")
        return f"wrote:{path}"

    def read_message(path: str) -> str:
        p = _safe_join(workspace, path)
        if not p.exists():
            return "not_found"
        return p.read_text(encoding="utf-8")

    registry.register("write_file", write_message)
    registry.register("read_file", read_message)
    return registry


def main() -> None:
    workspace = Path(__file__).resolve().parent / "sandbox"
    registry = build_registry(workspace)
    print(registry.dispatch("write_file", path="demo.txt", content="hello"))
    print(registry.dispatch("read_file", path="demo.txt"))


if __name__ == "__main__":
    main()

File: chapters/02_channel_adapter_registry/v3_registry_dispatch.ts
// TypeScript mirror for Chapter 02: Channel Adapter Registry (v3)
// Run with: npx ts-node v3_registry_dispatch.ts

import * as fs from "fs";
import * as path from "path";

class AdapterRegistry {
  private workspace: string;
  private handlers: Map<string, (...args: any[]) => string>;

  constructor(workspace: string) {
    this.workspace = path.resolve(workspace);
    if (!fs.existsSync(this.workspace)) {
      fs.mkdirSync(this.workspace, { recursive: true });
    }
    this.handlers = new Map();
  }

  register(name: string, handler: (...args: any[]) => string): void {
    this.handlers.set(name, handler);
  }

  dispatch(name: string, ...args: any[]): string {
    const handler = this.handlers.get(name);
    if (!handler) return `unknown adapter:${name}`;
    return handler(...args);
  }
}

function safeJoin(base: string, relPath: string): string {
  const baseResolved = path.resolve(base);
  const candidate = path.resolve(baseResolved, relPath);
  if (!candidate.startsWith(baseResolved)) {
    throw new Error("path escape rejected");
  }
  return candidate;
}

function buildRegistry(workspace: string): AdapterRegistry {
  const registry = new AdapterRegistry(workspace);

  registry.register("write_file", (filePath: string, content: string): string => {
    const p = safeJoin(workspace, filePath);
    const dir = path.dirname(p);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    fs.writeFileSync(p, content, "utf-8");
    return `wrote:${filePath}`;
  });

  registry.register("read_file", (filePath: string): string => {
    const p = safeJoin(workspace, filePath);
    if (!fs.existsSync(p)) return "not_found";
    return fs.readFileSync(p, "utf-8");
  });

  return registry;
}

function main(): void {
  const workspace = path.join(__dirname, "sandbox");
  const registry = buildRegistry(workspace);
  console.log(registry.dispatch("write_file", "demo.txt", "hello"));
  console.log(registry.dispatch("read_file", "demo.txt"));
}

main();
